[enhancement] Allow `if` without `else` in DSLX

### What's hard to do? (limit 100 words)

Currently, you can't have an `if` without an `else` part, which is different from Rust, where an `if` without an `else` always returns a unit type:

```
error[E0317]: `if` may be missing an `else` clause
 --> src/main.rs:2:5
  |
2 | /     if true {
3 | |         3
  | |         - found here
4 | |     };
  | |_____^ expected integer, found `()`
  |
  = note: `if` expressions without `else` evaluate to `()`
  = help: consider adding an `else` block that evaluates to the expected type
```

This is a small addition with well-defined behavior that can move DSLX closer to Rust syntax.
The change is not crucial by any means. We usually write empty `else` blocks when printing data with `trace_fmt!`.

### Current best alternative workaround (limit 100 words)

Write `if` with an empty `else`

### Your view of the "best case XLS enhancement" (limit 100 words)

As stated before, this is not a crucial addition to the language, but in my opinion, it could make code more readable. One thing to consider is the usage of IO operations in such statements, but this IO operations would have to ignore the returned token. Once again, this should be well-defined behavior.

--- a/xls/dslx/diagnostics/maybe_explain_error.cc
+++ b/xls/dslx/diagnostics/maybe_explain_error.cc
@@ -110,6 +110,19 @@ absl::Status ExplainIfBlockWithTrailingSemi(const TypeMismatchErrorData& data,
                             file_table);
 }
 
+absl::Status ExplainIfConditionalBlockWithoutElse(
+    const TypeMismatchErrorData& data, const Conditional* conditional,
+    const FileTable& file_table) {
+  const StatementBlock* culprit =
+      dynamic_cast<const StatementBlock*>(data.lhs_node);
+  std::string message = absl::StrFormat(
+      "%s; note that if expression without else from @ %s evaluates to (). "
+      "Consider adding an else block that evaluates to the expected type.",
+      data.message, culprit->span().ToString(file_table));
+  return XlsTypeErrorStatus(data.error_span, *data.lhs, *data.rhs, message,
+                            file_table);
+}
+
 absl::Status ExplainIfConditionalBlockWithTrailingSemi(
     const TypeMismatchErrorData& data, const Conditional* conditional,
     bool lhs_is_unit, bool rhs_is_unit, const FileTable& file_table) {
@@ -173,8 +186,13 @@ absl::Status MaybeExplainError(const TypeMismatchErrorData& data,
       is_block_in_conditional(data.rhs_node)) {
     if (auto* conditional = get_conditional(data.lhs_node);
         conditional == get_conditional(data.rhs_node)) {
-      return ExplainIfConditionalBlockWithTrailingSemi(
-          data, conditional, lhs_is_unit, rhs_is_unit, file_table);
+      if (!conditional->HasElse() && !lhs_is_unit) {
+        return ExplainIfConditionalBlockWithoutElse(data, conditional,
+                                                    file_table);
+      } else {
+        return ExplainIfConditionalBlockWithTrailingSemi(
+            data, conditional, lhs_is_unit, rhs_is_unit, file_table);
+      }
     }
   }
 

--- a/xls/dslx/diagnostics/maybe_explain_error_test.cc
+++ b/xls/dslx/diagnostics/maybe_explain_error_test.cc
@@ -110,5 +110,19 @@ TEST(MaybeExplainErrorTest,
       "note that conditional block @ test.x:2:26-4:6 had a trailing semicolon");
 }
 
+TEST(MaybeExplainErrorTest,
+     ExplainIfConditionalWithoutElseReturnsMismatchingType) {
+  constexpr std::string_view kProgram = R"(fn f() {
+    if true {
+        u32:42
+    }
+}
+)";
+  TypecheckAndExplain(kProgram,
+                      "note that if expression without else from @ "
+                      "test.x:2:13-4:6 evaluates to (). Consider adding an "
+                      "else block that evaluates to the expected type");
+}
+
 }  // namespace
 }  // namespace xls::dslx

--- a/xls/dslx/fmt/ast_fmt.cc
+++ b/xls/dslx/fmt/ast_fmt.cc
@@ -1611,10 +1611,12 @@ DocRef FmtConditionalMultiline(const Conditional& n, Comments& comments,
       FmtBlock(*n.consequent(), comments, arena, /*add_curls=*/false),
       arena.hard_line()};
 
+  bool has_else = n.HasElse();
   std::variant<StatementBlock*, Conditional*> alternate = n.alternate();
   while (std::holds_alternative<Conditional*>(alternate)) {
     Conditional* elseif = std::get<Conditional*>(alternate);
     alternate = elseif->alternate();
+    has_else = elseif->HasElse();
     pieces.push_back(arena.ccurl());
     pieces.push_back(arena.space());
     pieces.push_back(arena.Make(Keyword::kElse));
@@ -1626,17 +1628,19 @@ DocRef FmtConditionalMultiline(const Conditional& n, Comments& comments,
     pieces.push_back(arena.hard_line());
   }
 
-  CHECK(std::holds_alternative<StatementBlock*>(alternate));
-
-  StatementBlock* else_block = std::get<StatementBlock*>(alternate);
-  pieces.push_back(arena.ccurl());
-  pieces.push_back(arena.space());
-  pieces.push_back(arena.Make(Keyword::kElse));
-  pieces.push_back(arena.space());
-  pieces.push_back(arena.ocurl());
-  pieces.push_back(arena.hard_line());
-  pieces.push_back(FmtBlock(*else_block, comments, arena, /*add_curls=*/false));
-  pieces.push_back(arena.hard_line());
+  if (has_else) {
+    CHECK(std::holds_alternative<StatementBlock*>(alternate));
+    StatementBlock* else_block = std::get<StatementBlock*>(alternate);
+    pieces.push_back(arena.ccurl());
+    pieces.push_back(arena.space());
+    pieces.push_back(arena.Make(Keyword::kElse));
+    pieces.push_back(arena.space());
+    pieces.push_back(arena.ocurl());
+    pieces.push_back(arena.hard_line());
+    pieces.push_back(
+        FmtBlock(*else_block, comments, arena, /*add_curls=*/false));
+    pieces.push_back(arena.hard_line());
+  }
   pieces.push_back(arena.ccurl());
 
   return ConcatNGroup(arena, pieces);
@@ -1657,16 +1661,20 @@ DocRef Fmt(const Conditional& n, Comments& comments, DocArena& arena) {
       arena.break1(),
   };
 
-  CHECK(std::holds_alternative<StatementBlock*>(n.alternate()));
-  const StatementBlock* else_block = std::get<StatementBlock*>(n.alternate());
-  pieces.push_back(arena.ccurl());
-  pieces.push_back(arena.space());
-  pieces.push_back(arena.Make(Keyword::kElse));
-  pieces.push_back(arena.space());
-  pieces.push_back(arena.ocurl());
-  pieces.push_back(arena.break1());
-  pieces.push_back(FmtBlock(*else_block, comments, arena, /*add_curls=*/false));
-  pieces.push_back(arena.break1());
+  if (n.HasElse()) {
+    CHECK(std::holds_alternative<StatementBlock*>(n.alternate()));
+    const StatementBlock* else_block = std::get<StatementBlock*>(n.alternate());
+    pieces.push_back(arena.ccurl());
+    pieces.push_back(arena.space());
+    pieces.push_back(arena.Make(Keyword::kElse));
+    pieces.push_back(arena.space());
+    pieces.push_back(arena.ocurl());
+    pieces.push_back(arena.break1());
+    pieces.push_back(
+        FmtBlock(*else_block, comments, arena, /*add_curls=*/false));
+    pieces.push_back(arena.break1());
+  }
+
   pieces.push_back(arena.ccurl());
   return ConcatNGroup(arena, pieces);
 }

--- a/xls/dslx/fmt/ast_fmt_test.cc
+++ b/xls/dslx/fmt/ast_fmt_test.cc
@@ -503,6 +503,44 @@ TEST_F(FunctionFmtTest, ConditionalWithElseIfAfterLet) {
   EXPECT_EQ(got, original);
 }
 
+TEST_F(FunctionFmtTest, ConditionalWithoutElseShort) {
+  const std::string_view original =
+      R"(fn f(x:bool){if x {trace_fmt!("Hello DSLX");};})";
+  XLS_ASSERT_OK_AND_ASSIGN(std::string got, DoFmt(original, {"trace_fmt!"}));
+  const std::string_view want =
+      R"(fn f(x: bool) { if x { trace_fmt!("Hello DSLX"); }; })";
+  EXPECT_EQ(got, want);
+}
+
+TEST_F(FunctionFmtTest, ConditionalWithoutElseMultiStatementCausesHardBreaks) {
+  const std::string_view original =
+      R"(fn f(x:bool){if x{trace_fmt!("Hello");trace_fmt!("DSLX");};})";
+  XLS_ASSERT_OK_AND_ASSIGN(std::string got, DoFmt(original, {"trace_fmt!"}));
+  const std::string_view want =
+      R"(fn f(x: bool) {
+    if x {
+        trace_fmt!("Hello");
+        trace_fmt!("DSLX");
+    };
+})";
+  EXPECT_EQ(got, want);
+}
+
+TEST_F(FunctionFmtTest, ConditionalWithElseIfButWithoutElse) {
+  const std::string_view original =
+      R"(fn f(a:bool[2]){if a[0]{trace_fmt!("A0");}else if a[1]{trace_fmt!("A1");};})";
+  XLS_ASSERT_OK_AND_ASSIGN(std::string got, DoFmt(original, {"trace_fmt!"}));
+  const std::string_view want =
+      R"(fn f(a: bool[2]) {
+    if a[0] {
+        trace_fmt!("A0");
+    } else if a[1] {
+        trace_fmt!("A1");
+    };
+})";
+  EXPECT_EQ(got, want);
+}
+
 TEST_F(FunctionFmtTest, ConditionalWithUnnecessaryParens) {
   const std::string_view original =
       "fn f(a:u32,b:u32)->u32{if(a<b){a}else if(b<a){b}else{a}}";
@@ -2700,6 +2738,23 @@ TEST_F(ModuleFmtTest, IfElseIf) {
     } else {
         u32:42
     }
+}
+)");
+}
+
+TEST_F(ModuleFmtTest, IfWithoutElse) {
+  DoFmt(
+      R"(fn elseif_without_else(s: bool, x: u32) { if s == true { trace_fmt!("{}", x); }; }
+)");
+}
+
+TEST_F(ModuleFmtTest, IfElseIfWithoutElse) {
+  DoFmt(R"(fn elseif_without_else(s: bool, x: u32, y: u32) {
+    if s == true {
+        trace_fmt!("{}", x);
+    } else if x == u32:7 {
+        trace_fmt!("{}", y);
+    };
 }
 )");
 }

--- a/xls/dslx/frontend/ast.cc
+++ b/xls/dslx/frontend/ast.cc
@@ -640,11 +640,12 @@ NameDef::~NameDef() = default;
 Conditional::Conditional(Module* owner, Span span, Expr* test,
                          StatementBlock* consequent,
                          std::variant<StatementBlock*, Conditional*> alternate,
-                         bool in_parens)
+                         bool in_parens, bool has_else)
     : Expr(owner, std::move(span), in_parens),
       test_(test),
       consequent_(consequent),
-      alternate_(alternate) {}
+      alternate_(alternate),
+      has_else_(has_else) {}
 
 Conditional::~Conditional() = default;
 
@@ -665,14 +666,21 @@ std::vector<StatementBlock*> Conditional::GatherBlocks() {
 
 std::string Conditional::ToStringInternal() const {
   std::string inline_str = absl::StrFormat(
-      R"(if %s %s else %s)", test_->ToInlineString(),
-      consequent_->ToInlineString(), ToAstNode(alternate_)->ToInlineString());
+      R"(if %s %s%s)", test_->ToInlineString(), consequent_->ToInlineString(),
+      has_else_
+          ? absl::StrFormat(" else %s", ToAstNode(alternate_)->ToInlineString())
+          : std::string());
+
   if (inline_str.size() <= kTargetLineChars) {
     return inline_str;
   }
-  return absl::StrFormat(R"(if %s %s else %s)", test_->ToString(),
-                         consequent_->ToString(),
-                         ToAstNode(alternate_)->ToString());
+
+  return absl::StrFormat(
+      R"(if %s %s%s)", test_->ToString(), consequent_->ToString(),
+      has_else_ ? absl::StrFormat(" else %s", ToAstNode(alternate_)->ToString())
+                : std::string()
+
+  );
 }
 
 bool Conditional::HasMultiStatementBlocks() const {

--- a/xls/dslx/frontend/ast.h
+++ b/xls/dslx/frontend/ast.h
@@ -2069,7 +2069,7 @@ class Conditional : public Expr {
  public:
   Conditional(Module* owner, Span span, Expr* test, StatementBlock* consequent,
               std::variant<StatementBlock*, Conditional*> alternate,
-              bool in_parens = false);
+              bool in_parens = false, bool has_else = true);
 
   ~Conditional() override;
 
@@ -2097,10 +2097,11 @@ class Conditional : public Expr {
     return alternate_;
   }
 
+  bool HasElse() const { return has_else_; }
+
   bool HasElseIf() const {
     return std::holds_alternative<Conditional*>(alternate());
   }
-
   // Returns whether the blocks inside of this (potentially laddered)
   // conditional have multiple statements.
   bool HasMultiStatementBlocks() const;
@@ -2119,6 +2120,7 @@ class Conditional : public Expr {
   Expr* test_;
   StatementBlock* consequent_;
   std::variant<StatementBlock*, Conditional*> alternate_;
+  bool has_else_;
 };
 
 // Represents a member in a parametric binding list.

--- a/xls/dslx/frontend/ast_cloner.cc
+++ b/xls/dslx/frontend/ast_cloner.cc
@@ -921,7 +921,7 @@ class AstCloner : public AstNodeVisitor {
     old_to_new_[n] = module_->Make<Conditional>(
         n->span(), down_cast<Expr*>(old_to_new_.at(n->test())),
         down_cast<StatementBlock*>(old_to_new_.at(n->consequent())),
-        new_alternate, n->in_parens());
+        new_alternate, n->in_parens(), n->HasElse());
     return absl::OkStatus();
   }
 

--- a/xls/dslx/frontend/parser.cc
+++ b/xls/dslx/frontend/parser.cc
@@ -901,20 +901,28 @@ absl::StatusOr<Conditional*> Parser::ParseConditionalNode(
                       MakeRestrictions({ExprRestriction::kNoStructLiteral})));
   XLS_ASSIGN_OR_RETURN(StatementBlock * consequent,
                        ParseBlockExpression(bindings));
-  XLS_RETURN_IF_ERROR(DropKeywordOrError(Keyword::kElse));
+
+  XLS_ASSIGN_OR_RETURN(bool has_else, PeekTokenIs(Keyword::kElse));
 
   std::variant<StatementBlock*, Conditional*> alternate;
-
-  XLS_ASSIGN_OR_RETURN(const Token* peek, PeekToken());
-  if (peek->IsKeyword(Keyword::kIf)) {  // else if
-    XLS_ASSIGN_OR_RETURN(alternate,
-                         ParseConditionalNode(bindings, kNoRestrictions));
-  } else {
-    XLS_ASSIGN_OR_RETURN(alternate, ParseBlockExpression(bindings));
+  if (has_else) {
+    XLS_RETURN_IF_ERROR(DropKeywordOrError(Keyword::kElse));
+    XLS_ASSIGN_OR_RETURN(const Token* peek, PeekToken());
+    if (peek->IsKeyword(Keyword::kIf)) {  // else if
+      XLS_ASSIGN_OR_RETURN(alternate,
+                           ParseConditionalNode(bindings, kNoRestrictions));
+    } else {  // normal else
+      XLS_ASSIGN_OR_RETURN(alternate, ParseBlockExpression(bindings));
+    }
+  } else {  // if without else, add artificial empty block
+    alternate = module_->Make<StatementBlock>(Span(GetPos(), GetPos()),
+                                              std::vector<Statement*>(),
+                                              /*trailing_semi=*/true);
   }
 
   auto* outer_conditional = module_->Make<Conditional>(
-      Span(if_kw.span().start(), GetPos()), test, consequent, alternate);
+      Span(if_kw.span().start(), GetPos()), test, consequent, alternate,
+      /*in_parens=*/false, has_else);
   for (StatementBlock* block : outer_conditional->GatherBlocks()) {
     block->SetEnclosing(outer_conditional);
   }

--- a/xls/dslx/frontend/parser_test.cc
+++ b/xls/dslx/frontend/parser_test.cc
@@ -2564,6 +2564,8 @@ TEST_F(ParserTest, ForFreevars) {
                              absl::flat_hash_set<std::string>{"j", "range"}));
 }
 
+TEST_F(ParserTest, IfWithoutElse) { RoundTripExpr("if true {}"); }
+
 TEST_F(ParserTest, EmptyTernary) { RoundTripExpr("if true {} else {}"); }
 
 TEST_F(ParserTest, TernaryConditional) {

--- a/xls/dslx/tests/errors/error_modules_test.py
+++ b/xls/dslx/tests/errors/error_modules_test.py
@@ -1311,6 +1311,16 @@ class ImportModuleWithTypeErrorTest(test_base.TestCase):
         stderr,
     )
 
+  def test_if_without_else_returning_mismatching_type(self):
+    stderr = self._run(
+        "xls/dslx/tests/errors/if_without_else_returning_mismatching_type.x"
+    )
+    self.assertIn("XlsTypeError:", stderr)
+    self.assertIn(
+        "uN[32] vs (): Conditional consequent type (in the 'then' clause) did "
+        "not match alternative type (in the 'else' clause)",
+        stderr,
+    )
 
 if __name__ == '__main__':
   test_base.main()

--- a/xls/dslx/tests/errors/if_without_else_returning_mismatching_type.x
+++ b/xls/dslx/tests/errors/if_without_else_returning_mismatching_type.x
@@ -1,0 +1,17 @@
+// Copyright 2025 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+fn f(predicate: bool) {
+  if predicate { u32:0 }
+}

